src/clib/initialize_cloudy_data.cpp

```diff
@@ -0,0 +1,290 @@
+  if (!h5io::GridTableProps_is_valid(grid_props)) {
+    h5io::drop_GridTableProps(&grid_props);
+    H5Fclose (file_id);
+    // error messages were already printed by h5io::parse_GridTableProps
+    return GR_FAIL;
+  }
+
+  // use grid_props to initialize parts of my_cloudy
+  // TODO: it would be really nice if we explicitly validated that each
+  //       grid_props.axes[i] held the expected values
+  my_cloudy->grid_rank = static_cast<long long>(grid_props.table_shape.ndim);
+  for (int i = 0; i < grid_props.table_shape.ndim; i++) {
+    my_cloudy->grid_dimension[i]
+      = static_cast<long long>(grid_props.table_shape.shape[i]);
+
+    bool is_temperature =
+      std::strcmp("Temperature", grid_props.axes[i].name) == 0;
+
+    double* buf = new double[my_cloudy->grid_dimension[i]];
+    for (long long w = 0LL; w < my_cloudy->grid_dimension[i]; w++) {
+      if (is_temperature) {
+        buf[w] = std::log10(grid_props.axes[i].values[w]);
+      } else {
+        buf[w] = grid_props.axes[i].values[w];
+      }
+    }
+    my_cloudy->grid_parameters[i] = buf;
+  }
+
+  if (grackle_verbose) {
+    std::fprintf(stdout, "Cloudy cooling grid = {\n");
+    std::fprintf(stdout, "  rank: %lld,\n", my_cloudy->grid_rank);
+    for (long long i = 0LL; i < my_cloudy->grid_rank; i++) {
+      std::fprintf(stdout,
+          "  axis %lld (\"%s\"): %g to %g (%lld steps),\n",
+          i,
+          grid_props.axes[i].name,
+          my_cloudy->grid_parameters[i][0],
+          my_cloudy->grid_parameters[i][my_cloudy->grid_dimension[i]-1],
+          my_cloudy->grid_dimension[i]);
+    }
+    fprintf(stdout, "}\n");
+  }
+  h5io::ArrayShape expected_shape = grid_props.table_shape;
+
+  // Read Cooling data.
+  my_cloudy->data_size = h5io::ArrayShape_elem_count(expected_shape);
+  my_cloudy->cooling_data = load_heatcool_data(file_id, dset_name, CoolUnit,
+                                               expected_shape);
+
+  if (my_cloudy->cooling_data == nullptr) {
+    h5io::drop_GridTableProps(&grid_props);
+    H5Fclose(file_id);
+    return GR_FAIL;
+  }
+
+  // Read Heating data.
+  if (my_chemistry->UVbackground == 1) {
+    std::snprintf(dset_name, name_bufsize, "/CoolingRates/%s/Heating",
+                  group_name);
+
+    // Ideally we would uncomment the following block of logic that verifies
+    // that the Grid Table properties are identical to the cooling table
+    // -> uncommenting the logic would be useful for enforcing consistency
+    //    in any new data files (all files currently shipped with Grackle
+    //    already satisfy this requirement)
+    // -> the only downside to uncommenting the block is a little overhead
+
+    //if (h5io::assert_has_consistent_GridTableProps(file_id, dset_name,
+    //                                               grid_props) != GR_SUCCESS){
+    //  h5io::drop_GridTableProps(&grid_props);
+    //  H5Fclose(file_id);
+    //  return GR_FAIL;
+    //}
```

Comment:
I would really like to make this change. The only drawback is that it will introduce slight overhead at startup. But, I think it's probably worth it

```suggestion
    // validate that Heating table has have identical GridTableProps
    if (h5io::assert_has_consistent_GridTableProps(file_id, dset_name,
                                                   grid_props) != GR_SUCCESS){
      h5io::drop_GridTableProps(&grid_props);
      H5Fclose(file_id);
      return GR_FAIL;
    }
```